File: String/Minimum Penalty for a Shop.cpp
class Solution {
public:
    int bestClosingTime(string customers) {
        //brute force

        int minPenalty = INT_MAX;
        int minHour = INT_MAX;
        int n = customers.size();

        for(int i = 0;i<=n;i++){
            int penalty = 0;
            int j = i-1;
            while(i > 0 && j >= 0){
                if(customers[j] == 'N'){
                    penalty++;
                }
                j--;
            }
            j = i;
            while(i < n && j < n){
                if(customers[j] == 'Y'){
                    penalty++;
                }
                j++;
            }
            if(minPenalty > penalty){
                minPenalty = penalty;
                minHour = i;
            }
        }

        return minHour;

        //optimal

        int minHour = 0;
        int penalty = count(begin(customers), end(customers), 'Y');
        int minPenalty = penalty;
        int n = customers.size();

        for(int i = 0;i<n;i++){
            if(customers[i] == 'Y'){
                penalty--;
            }else{
                penalty++;
            }
            if(minPenalty > penalty){
                minPenalty = penalty;
                minHour = i+1;
            }
        }
        return minHour;
    }
};

File: String/Minimum Penalty for a Shop.java
class Solution {
    public int bestClosingTime(String customers) {
      //better
        int minPenalty = Integer.MAX_VALUE;
        int minHour = Integer.MAX_VALUE;
        int n = customers.length();

        int []prefixN = new int[n+1];
        int []suffixY = new int[n+1];

        prefixN[0] = 0;
        suffixY[n] = 0;
        int sumN = 0;
        int sumY = 0;

        for(int i = 1;i<=n;i++){
            if(customers.charAt(i-1) == 'N'){
                sumN++;
            }
            prefixN[i] = sumN;
        }

        for(int i = n-1;i>=0;i--){
            if(customers.charAt(i) == 'Y'){
                sumY++;
            }
            suffixY[i] = sumY;
        }

        // System.out.println(Arrays.toString(prefixN));
        // System.out.println(Arrays.toString(suffixY));

        for(int i = 0;i<=n;i++){
            if(minPenalty > prefixN[i] + suffixY[i]){
                minPenalty = prefixN[i] + suffixY[i];
                minHour = i;
            }
        }

        return minHour;
    }
}
